### PR TITLE
DEVPROD-4502 Add task tags linting

### PR DESCRIPTION
**Pull request by MikhailShchatko**
Problem
-----
As part of `evergreen.yml rework` project we would like to introduce task tags linting, so that we would be able to enforce having `assigned_to_jira_team_*` tags on every task.

Solution
-----
Make `evergreen-ci/config-linter` great again.

A while ago we disabled and consequently removed `evergreen-ci/config-linter` from `mongodb-mongo-master` project. Now it's time to enhance and bring it back.

In this PR a new rule `enforce-tags-for-tasks` is implemented. With this it would be possible to configure tag groups with the following configuration:

```yaml
    - rule: "enforce-tags-for-tasks"
      tag_groups:
        - group_name: "tag_group_name"
          # the rule will fail, if tag matches the regex and do not match any tag from `tag_list`
          tag_regex: required_tag_.+
          tag_list:
            - required_tag_test_1
            - required_tag_test_2
            - required_tag_test_3
          # the rule will fail, if the number of tags is less than `min_num_of_tags` or more than `max_num_of_tags`
          min_num_of_tags: 1
          max_num_of_tags: 1
```

Testing
-----

1. Unit tests added in this PR
2. Ran linter against evergreen yml in the mongo repo
    - configuration:
    ```yaml
        - rule: "enforce-tags-for-tasks"
          tag_groups:
            - group_name: "team_assignment_tags"
              tag_regex: assigned_to_jira_team_.+
              tag_list:
                - assigned_to_jira_team_server_servicearch
                - assigned_to_jira_team_server_security
                - assigned_to_jira_team_server_cluster_scalability
                - assigned_to_jira_team_server_catalog_and_routing
                - assigned_to_jira_team_server_repl
                - assigned_to_jira_team_server_execution
                - assigned_to_jira_team_server_storage_engines
                - assigned_to_jira_team_server_query_execution
                - assigned_to_jira_team_server_query_optimization
                - assigned_to_jira_team_server_query_integration
                - assigned_to_jira_team_devprod_build
                - assigned_to_jira_team_devprod_correctness
                - assigned_to_jira_team_atlas_streams
              min_num_of_tags: 1
              max_num_of_tags: 1
    ```
    - ran commands:
    ```bash
    evergreen evaluate <mongo_repo>/etc/evergreen.yml > <mongo_repo>/etc/evaluated_evergreen.yml
    poetry run main -c <mongo_repo>/etc/evergreen_lint.yml lint
    ```
    - got results in ~5-10 seconds
    ```bash
    Load config file: <mongo_repo>/etc/evergreen_lint.yml
    6 errors found in '<mongo_repo>/etc/evaluated_evergreen.yml':
    For help resolving errors, see the helpful documentation at https://wiki.corp.mongodb.com/pages/viewpage.action?pageId=133273894
    enforce-tags-for-tasks: Task tags requirement is not met. The task 'sharded_collections_jscore_passthrough_with_replica_set_endpoint_gen' should have no less than '1' and no more than '1' tag(s) of 'team_assignment_tags' tag group list: ['assigned_to_jira_team_server_servicearch', 'assigned_to_jira_team_server_security', 'assigned_to_jira_team_server_cluster_scalability', 'assigned_to_jira_team_server_catalog_and_routing', 'assigned_to_jira_team_server_repl', 'assigned_to_jira_team_server_execution', 'assigned_to_jira_team_server_storage_engines', 'assigned_to_jira_team_server_query_execution', 'assigned_to_jira_team_server_query_optimization', 'assigned_to_jira_team_server_query_integration', 'assigned_to_jira_team_devprod_build', 'assigned_to_jira_team_devprod_correctness', 'assigned_to_jira_team_atlas_streams']. Found tags: []. Please add/remove tag(s) of the tag group list to/from the task 'sharded_collections_jscore_passthrough_with_replica_set_endpoint_gen' to match the requirement.
    
    enforce-tags-for-tasks: Task tags requirement is not met. The task 'sharded_jscore_txns_with_replica_set_endpoint_gen' should have no less than '1' and no more than '1' tag(s) of 'team_assignment_tags' tag group list: ['assigned_to_jira_team_server_servicearch', 'assigned_to_jira_team_server_security', 'assigned_to_jira_team_server_cluster_scalability', 'assigned_to_jira_team_server_catalog_and_routing', 'assigned_to_jira_team_server_repl', 'assigned_to_jira_team_server_execution', 'assigned_to_jira_team_server_storage_engines', 'assigned_to_jira_team_server_query_execution', 'assigned_to_jira_team_server_query_optimization', 'assigned_to_jira_team_server_query_integration', 'assigned_to_jira_team_devprod_build', 'assigned_to_jira_team_devprod_correctness', 'assigned_to_jira_team_atlas_streams']. Found tags: []. Please add/remove tag(s) of the tag group list to/from the task 'sharded_jscore_txns_with_replica_set_endpoint_gen' to match the requirement.
    
    enforce-tags-for-tasks: Task tags requirement is not met. The task 'sharded_jscore_txns_sharded_collections_with_replica_set_endpoint_gen' should have no less than '1' and no more than '1' tag(s) of 'team_assignment_tags' tag group list: ['assigned_to_jira_team_server_servicearch', 'assigned_to_jira_team_server_security', 'assigned_to_jira_team_server_cluster_scalability', 'assigned_to_jira_team_server_catalog_and_routing', 'assigned_to_jira_team_server_repl', 'assigned_to_jira_team_server_execution', 'assigned_to_jira_team_server_storage_engines', 'assigned_to_jira_team_server_query_execution', 'assigned_to_jira_team_server_query_optimization', 'assigned_to_jira_team_server_query_integration', 'assigned_to_jira_team_devprod_build', 'assigned_to_jira_team_devprod_correctness', 'assigned_to_jira_team_atlas_streams']. Found tags: []. Please add/remove tag(s) of the tag group list to/from the task 'sharded_jscore_txns_sharded_collections_with_replica_set_endpoint_gen' to match the requirement.
    
    enforce-tags-for-tasks: Task tags requirement is not met. The task 'sharding_jscore_passthrough_with_replica_set_endpoint_gen' should have no less than '1' and no more than '1' tag(s) of 'team_assignment_tags' tag group list: ['assigned_to_jira_team_server_servicearch', 'assigned_to_jira_team_server_security', 'assigned_to_jira_team_server_cluster_scalability', 'assigned_to_jira_team_server_catalog_and_routing', 'assigned_to_jira_team_server_repl', 'assigned_to_jira_team_server_execution', 'assigned_to_jira_team_server_storage_engines', 'assigned_to_jira_team_server_query_execution', 'assigned_to_jira_team_server_query_optimization', 'assigned_to_jira_team_server_query_integration', 'assigned_to_jira_team_devprod_build', 'assigned_to_jira_team_devprod_correctness', 'assigned_to_jira_team_atlas_streams']. Found tags: []. Please add/remove tag(s) of the tag group list to/from the task 'sharding_jscore_passthrough_with_replica_set_endpoint_gen' to match the requirement.
    
    enforce-tags-for-tasks: Task tags requirement is not met. The task 'disk_wiredtiger' has tags that matches 'assigned_to_jira_team_.+' tag regex of 'team_assignment_tags' tag group and is not listed in the group: ['assigned_to_jira_team_server_servicearch', 'assigned_to_jira_team_server_security', 'assigned_to_jira_team_server_cluster_scalability', 'assigned_to_jira_team_server_catalog_and_routing', 'assigned_to_jira_team_server_repl', 'assigned_to_jira_team_server_execution', 'assigned_to_jira_team_server_storage_engines', 'assigned_to_jira_team_server_query_execution', 'assigned_to_jira_team_server_query_optimization', 'assigned_to_jira_team_server_query_integration', 'assigned_to_jira_team_devprod_build', 'assigned_to_jira_team_devprod_correctness', 'assigned_to_jira_team_atlas_streams']. Please remove the following tag(s) from the task 'disk_wiredtiger': ['assigned_to_jira_team_test'].
    
    enforce-tags-for-tasks: Task tags requirement is not met. The task 'disk_wiredtiger' should have no less than '1' and no more than '1' tag(s) of 'team_assignment_tags' tag group list: ['assigned_to_jira_team_server_servicearch', 'assigned_to_jira_team_server_security', 'assigned_to_jira_team_server_cluster_scalability', 'assigned_to_jira_team_server_catalog_and_routing', 'assigned_to_jira_team_server_repl', 'assigned_to_jira_team_server_execution', 'assigned_to_jira_team_server_storage_engines', 'assigned_to_jira_team_server_query_execution', 'assigned_to_jira_team_server_query_optimization', 'assigned_to_jira_team_server_query_integration', 'assigned_to_jira_team_devprod_build', 'assigned_to_jira_team_devprod_correctness', 'assigned_to_jira_team_atlas_streams']. Found tags: []. Please add/remove tag(s) of the tag group list to/from the task 'disk_wiredtiger' to match the requirement.
    ```

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,40 +1,41 @@
...
 # Python version should be set to the minimum supported toolchain version
-python = "^3.7"
+python = "^3.9"
 # These dependencies are tied to whatever versions are used by the oldest
 # supported branch of MongoDB server, presently 5.0
 PyYAML = ">=3.0.0"
-typing-extensions = "^3.10.0"
-click = "~=7.0"
+typing-extensions = ">=3.10.0"
+click = ">=7.0"
```
Comment (MikhailShchatko) [inline pyproject.toml:20]: I hope that changes above would make it easier to add this package as a dependency in the mongo repo.

File: pyproject.toml [hunk 2]
```diff
@@ -1,40 +1,41 @@
...
+pytest-black = "^0.3"
+pytest-flake8 = "^1.0"
+pytest-mypy = "^0.8"
+pytest-isort = "^2.0"
+types-PyYAML = "^6.0.11"
 
 [tool.black]
 line-length = 100
-target-version = ['py37']
+target-version = ['py39']
```
Comment (IamXander) [inline pyproject.toml:35]: We are fully on 3.10 now

Comment (IamXander) [reply pyproject.toml:35]: oh sorry, i thought this was a different repo, please ignore this comment

**Review by Trevor159 — APPROVED** [review]
LGTM, nice job

**Pull request merged**